scheduler/src/cook/queue_limit.clj

```diff
@@ -0,0 +1,155 @@
+      pool-name
+      field)
+    Integer/MAX_VALUE))
+
+(defn- pool-global-threshold
+  "Returns the pool-global-threshold for the given pool,
+  the value at which the per-user queue limit switches
+  from the 'normal' number to the 'constrained' number"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :pool-global-threshold))
+
+(defn- user-limit-normal
+  "Returns the user-limit-normal for the given pool"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :user-limit-normal))
+
+(defn- user-limit-constrained
+  "Returns the user-limit-constrained for the given pool"
+  [pool-name]
+  (match-based-on-pool-name
+    pool-name
+    :user-limit-constrained))
+
+(defn- update-interval-seconds
+  "Returns the interval in seconds at which
+  to refresh queue lengths from the database"
+  []
+  (:update-interval-seconds
+    (config/queue-limits)))
+
+(defn get-pending-jobs
+  "Queries for and returns the set of
+  currently pending jobs from the database"
+  []
+  (-> datomic/conn
+      d/db
+      queries/get-pending-job-ents))
+
+(defn query-queue-lengths
+  "Queries for pending jobs from the database and
+  returns a map with two sub-maps of the following shape:
+
+  {:pool->queue-length {pool-a 100 pool-b 200 ...}
+   :pool->user->queue-length {pool-a {user-x 10 user-y 20 user-z 70}
+                              pool-b {user-x 20 user-y 40 user-z 140}
+                              ...}"
+  []
+  (let [pending-jobs
+        (get-pending-jobs)
+        pool->pending-jobs
+        (group-by
+          caches/job->pool-name
+          pending-jobs)]
+    {:pool->queue-length
+     (pc/map-vals count pool->pending-jobs)
+     :pool->user->queue-length
+     (pc/map-vals
+       #(pc/map-vals
+          count
+          (group-by
+            caches/job-ent->user
+            %))
+       pool->pending-jobs)}))
+
+(let [pool->queue-length-atom (atom {})
+      pool->user->queue-length-atom (atom {})]
+
+  (defn user-queue-length
+    "Returns the queue length for the given pool name and user"
+    [pool-name user]
+    (get-in
+      @pool->user->queue-length-atom
+      [pool-name user]
+      0))
+
+  (defn user-queue-limit
+    "Returns the queue length limit for the given pool name -- if the
+    pool-global queue length is <= than the pool global threshold, we
+    use the 'normal' per-user limit, otherwise, we switch to using the
+    'constrained' per-user limit"
+    [pool-name]
+    (let [pool-global-length (get @pool->queue-length-atom pool-name 0)
+          pool-global-threshold (pool-global-threshold pool-name)]
+      (if (<= pool-global-length pool-global-threshold)
+        (user-limit-normal pool-name)
+        (user-limit-constrained pool-name))))
+
+  (defn inc-queue-length!
+    "Increments the pool-global and per-user queue lengths for
+    the given pool name and user by the given number of jobs"
+    [pool-name user number-jobs]
+    (let [inc-number-jobs #(-> % (or 0) (+ number-jobs))]
```

Comment:
Probably worth an assert that pool-name is not nil.

Reply:
Done